geoh5py/ui_json/forms.py

```diff
@@ -0,0 +1,499 @@
+            out = self.snake_to_camel.get(key, key)
+        else:
+            raise ValueError("Convention must be 'snake' or 'camel'.")
+
+        return out
+
+    def map(self, collection: dict[str, Any], convention="snake"):
+        """Map a dictionary from snake to camel or vice versa."""
+        return {self.map_key(k, convention): v for k, v in collection.items()}
+
+
+MEMBER_KEYS = MemberKeys()
+
+
+class ValueAccess:
+    """
+    Descriptor to elevate underlying member values within 'FormParameter'.
+
+    :param private: Name of private attribute.
+    """
+
+    def __init__(self, private: str):
+        self.private = private
+
+    def __get__(self, obj, objtype=None):
+        return getattr(obj, self.private).value
+
+    def __set__(self, obj, value):
+        setattr(getattr(obj, self.private), "value", value)
+        obj._active_members.append(self.private[1:])
+
+
+class FormParameter:
+    """
+    Base class for parameters that create visual ui elements from a form.
+
+    :param name: Parameter name.
+    :param label: Label for ui element.
+    :param value: The parameter's value.
+    :param enabled: If False, ui element is rendered grey and value is
+        treated as None.
+    :param optional: If True, ui element is rendered with a checkbox to
+        control the enabled state.
+    :param group_optional: If True, ui group is rendered with a checkbox
+        that control the enabled state of all group members.
+    :param main: Controls whether ui element will render in the general
+        parameters tab (True) or optional parameters (False).
+    :param group: Grouped ui elements will be rendered within a named
+        box.
+    :param dependency: Name of parameter that controls the enabled or
+        visible state of the ui element.
+    :param dependency_type: Controls whether the ui element is enabled
+        or visible when the dependency is enabled if optional or True
+        if a bool type.
+    :param group_dependency: Name of parameter that controls the enabled
+        or visible state of the ui group.
+    :param group_dependency_type: Controls whether the ui group is
+        enabled or visible when the group dependency is enabled if
+        optional or True if a bool type.
+    :param tooltip: String rendered on hover over ui element.
+
+    :note: Standardized form members are accessible through public namespace
+        by way of the ValueAccess descriptor.
+    """
+
+    validations = {"required_form_members": ["label", "value"]}
+    identifier_members: list[str] = []
+
+    def __init__(
+        self,
+        name: str,
+        value: Any | None = None,
+        **kwargs,
+    ):
+        self.name: str = name
+        self._value: Parameter = self._set_value_parameter(value)
+        self._label = StringParameter("label")
+        self._enabled = BoolParameter("enabled", value=True)
+        self._optional = BoolParameter("optional", value=False)

```

Comment:
See comment about default state of BoolParameter. Maybe we just specify only if changed from default?

Reply:
done.